### ISSUE TITLE
gnutls not enabled on macOS - "static libetpan" target needs gnutls

### ISSUE THREAD

**Issue by AlexanderAA** [dinhvh/libetpan#383]
Hello,

Would you be able to advise what is the correct way to build on macOS with SSL/TLS (gnutls) enabled? It looks like "static libetpan" target is built without gnutls at the moment. 

More details on why this is needed:
https://github.com/Homebrew/homebrew-core/issues/54182#issuecomment-633126238

**Comment by dwery** [comment]
FWIW this issue still exists

**Comment by j6s** [comment]
Seems like this issue currently still exists

**Comment by gith-account** [comment]
The problem with libetpan as described in homebrew-core issue [#54182](https://github.com/Homebrew/homebrew-core/issues/54182) still persists.

> Would you be able to advise what is the correct way to build on macOS with SSL/TLS (gnutls) enabled?

Unfortunately no response of any kind has been received here in 4 years.

Claws-mail 4.3.1 installs fine and can successfully establish TLS connections if the MacOS related instructions with regard to libetpan are removed from the claws-mail and libetpan homebrew formulas (see attachments).

@AlexanderAA, maybe ask @alebcay again to reopen homebrew-core issue [#54182](https://github.com/Homebrew/homebrew-core/issues/54182) and adapt the currently published formulas.

As long as there is no input from a qualified party on how it could and should be done better, a homebrew formula that does not follow libetpan recommendations but (after an initial assessment) results in a working claws-mail installation seems better than a definitely broken formula that cluelessly follows uncommented recommendations for a possibly better implementatiom, without success and  a working outcome.

[claws-mail.rb.txt](https://github.com/user-attachments/files/19090939/claws-mail.rb.txt)
[libetpan.rb.txt](https://github.com/user-attachments/files/19090938/libetpan.rb.txt)

**Comment by CL-Jeremy** [comment]
@gith-account Thanks for digging into this. Some thoughts:

* For this library, it does not matter what SSL library is chosen for Claws-Mail, so the current Homebrew library choice can be kept (tried myself with either OpenSSL or GnuTLS from Homebrew, both work with Claws-Mail linked to GnuTLS only)
* Linking to GnuTLS fails with manually specifying a prefix (presumably due to a bug in Autoconf), while `--with-gnutls=yes` works perfectly (with Homebrew shims, this suffices to make Autoconf check in `opt_prefix`)
* The correct option for specifying SASL2 path is `--with-sasl` (without the 2), but leaving this out makes no difference
* Basically autodetection works fine with the current choice of library in Homebrew formula, so all the `--with-*` could be dropped
* The official macOS targets (framework and static library) uses the system framework `CFNetwork` for TLS support, which is currently not working with Claws-Mail (confirmed tests in this repo running fine when built against `libetpan.framework`)
* The solution to this problem: either use the generic *nix build process with Autoconf and build a normal shared library (.dylib) against either OpenSSL or GnuTLS, or fix Claws-Mail to support `libetpan` built against `CFNetwork`
  - In the former case, `depends_on xcode: :build` could be dropped

There is also an open issue at #445 dealing with `CFNetwork` availability on older Mac OS X versions. AFAIK, MacPorts never bothered with building frameworks for this library, so PPCPorts users have no issues there.

**Comment by dinhvh** [comment]
Does CFNetwork works for this?

**Comment by CL-Jeremy** [comment]
> Does CFNetwork works for this?

No, since Claws Mail needs GnuTLS for its own functionalities regardless. The way libetpan handles TLS is too different when using CFNetwork instead of OpenSSL/GnuTLS that Claws Mail cannot work with it.

Basically this boils down to Homebrew not having options for its packages and is normally reluctant to package different variants of the same software. It is so simplistic that even its dependency spec does not support "or" (as in the functionality could be satisfied by any one of the specified packages). It is thus standard practice to include all functionalities compiled in, so as to cater to all needs from different softwares. Unfortunately, this currently breaks Claws Mail.

I believe it is correct for libetpan to prioritize CFNetwork when available. If one doesn't want to use an API, they simply shouldn't link to it. Homebrew needs to try and sort out the linking problem, especially as CFNetwork is not even intended when linking to libetpan dynamically. The Homebrew project is a giant infrastructure for open source software, and thus shouldn't be using CFNetwork anyway, especially in conjunction with OpenSSL which makes no sense.

**Comment by dinhvh** [comment]
@CL-Jeremy 
would you be able to check if the current master would allow to choose gnutls or openssl:
https://github.com/dinhvh/libetpan/commit/ee12b3b457a02eb9aa37ff9cf50ebfb5a124bcff

This commit would allow to build against CFNetwork + OpenSSL + GnuTLS
Then at runtime, we can decide which implementation we use.

I believe that would help resolve the problem from the perspective of the repo.

**Comment by CL-Jeremy** [comment]
Sorry, I should have noticed the changes on Homebrew's side, and saved you all this work.

I'm primarily working with MacPorts recently, which has always adhered to standard UNIX conventions for `libetpan`, whereas Homebrew chose to compile it like an iOS/macOS app dependency (framework + static archive, which is not even the intended use case for the target "static libetpan" – using that requires linking to CoreFoundation, which is not found as a dependency by its users, such as the Homebrew formula for Claws Mail, which had to override to using the dynamic framework).

Since libetpan 1.10.x, however, Homebrew maintainers have noticed the issue and switched to compiling it as a standard dylib with OpenSSL (would be the same with GnuTLS), which works fantastically with the other open-source packages, just like on Linux/FreeBSD. They did retain the CFNetwork linkage, but now with OpenSSL compiled in, Claws Mail works happily with it (it seems the necessity of the SSL backend was the certificate retrieval/check parts, not the actual network streams, so that might as well be handled by CFNetwork, Claws Mail does not care – both 1.10.1 and HEAD work flawlessly with libetpan built with `-DHAVE_CFNETWORK=1`, standard since 1.10 with Homebrew, but I suppose CFNetwork is only chosen by default on HEAD, since `LIBETPAN_IOS_DISABLE_SSL` is never set?).

I guess this change is still kind of useful, so the package maintainer don't have to manually modify the flags across OS versions and let the application developer choose whether to use the system framework backend, since it might not work properly on a very old OS? I personally don't believe packagers will make use of this functionality, so I have no incentive to use it, either. In any case, I'll do some further testing with different combinations of build flags and manual backend choices in the downstream application.

**Comment by dinhvh** [comment]
Do you think we can close this issue or is there anything that is needed from me here?

**Comment by CL-Jeremy** [comment]
True, this issue is solved downstream by the package maintainers and should be closed, as the last stable release was already functional for the original use case (which is *not exactly* what the title describes). Further discussions could be done elsewhere (I'll probably open an issue if there's something I cannot solve on myself).
